Add Environment Variable Override

trankin opened the pull request:
I need to be able to manage settings on this service via Environment Variables.   I've added the ability to override anything in the settings schema with a corresponding environment variable.  

e.g. to override settings.serial.port you can set ZIGBEE2MQTT_SERIAL_PORT=/dev/ttyS0
to override the mqtt username you can set ZIGBEE2MQTT_MQTT_USER=testusername

This new addition will not perform any action on existing installations unless the matching environment variable is set.  I have tested this setting string, number, boolean, object and array values.

trankin committed:
Add Environment Variable Override

I need to be able to manage settings on this service via Environment Variables.   I've added the ability to override anything in the settings schema with a corresponding environment variable.  

e.g. to override settings.serial.port you can set ZIGBEE2MQTT_SERIAL_PORT=/dev/ttyS0
to override the mqtt username you can set ZIGBEE2MQTT_MQTT_USER=testusername

This new addition will not perform any action on existing installations unless the matching environment variable is set.  I have tested this setting string, number, boolean, object and array values.

trankin commented:
I see this is failing for code coverage... I'm working on adding a test that will complete this.

trankin committed:
Adding test case for environment variables and slight modification to ensure 100% code coverage.

trankin committed:
Adding a test to confirm that env variables will set non default values.  Also realized that I was errantly applying the env variables to the defaults for testing.  Understanding what this is doing more clearly I realize that should be clean.

Koenkk commented:
Could you explain the use case for this?

trankin commented:
> Could you explain the use case for this?

Sure, basically it's because I'm using a fleet manager to deploy and manage applications on hosts.  I can set up configurations for all of the computers to share settings in the fleet manager via environment variables and it performs the work of syncing those settings down to the devices.   It's much easier for me to deploy changes across the entire fleet via environment variables rather than managing the configuration.yaml on a per device basis.

Koenkk commented on lib/util/settings.js:
```diff
@@ -129,6 +129,13 @@ const defaults = {
         timestamp_format: 'YYYY-MM-DD HH:mm:ss',
     },
     external_converters: [],
+    env_var_tests: {
```
I don't understand why test code ended up here, is this needed?

trankin replied:
I can refactor it to remove those from the settings.  Mostly just feeling my way around and realized that I needed to have some of those so that they show up in defaults for the existing tests.  I can mock those structures in the new tests that I've created and remove them from settings.js

Koenkk commented on lib/util/settings.js:
```diff
@@ -421,9 +439,52 @@ function read() {
+                if (key !== 'properties') {
+                    const type = (obj[key].type || 'object').toString();
+                    const envPart = path.reduce((acc, val) => `${acc}${val}_`, '');
+                    const envVariableName = (`ZIGBEE2MQTT_${envPart}${key}`).toUpperCase();
```
I propose to change the prefix from `ZIGBEE2MQTT_` to `ZIGBEE2MQTT_CONFIG_`

trankin replied:
I had a similar thought.  Will do.

trankin committed:
Refactoring to

1. Remove the test variables from the schema and defaults and manually reflect the tests in the test.

2. Rename environment variable base from ZIGBEE2MQTT_ to ZIGBEE2MQTT_CONFIG_

Koenkk committed:
Small improvements

Koenkk commented on test/settings.test.js:
```diff
@@ -53,6 +62,46 @@ describe('Settings', () => {
+        expect(s).toStrictEqual(expected);
+    });
+
+    it('Should apply environment variables and not match defaults', () => {
```
Looks good, last thing before merge.

I don't understand this testcase, I've also tried removing it and still the coverage is OK.

trankin replied:
It's really not relevant anymore now that I've moved those values out of defaults.   Can be safely removed.  Let me know if you want me to or if you will.

trankin replied:
disregard my last.. I'm removing that test now.

trankin committed:
Removing the unneeded test.

Koenkk commented:
Thanks, wrote some docs here: https://github.com/Koenkk/zigbee2mqtt.io/blob/develop/docs/information/configuration.md#override-via-environment-variables

qm3ster commented:
Can the overrides mutate configuration.yaml?
@trankin what do you do to manage `coordinator_backup.json` and `devices.yaml`(or corresponding section of `configuration.yaml` in your fleet?

trankin commented:
@qm3ster the overrides to not mutate configuration.yaml.  My addition of  the env variable feature simply overrides settings after the configuration.yaml file has been loaded.  In my set up devices.yaml and/or coordinator_backup.json would still be stored locally to the device.
